lib/reply.js

```diff
@@ -157,66 +142,91 @@ function onSendEnd (reply, payload) {
     if (!contentType) {
       reply.res.setHeader('Content-Type', 'application/octet-stream')
     }
-    pump(payload, reply.res, wrapPumpCallback(reply))
+    pump(payload, reply.res, pumpCallback(reply))
     return reply.res
   }
 
   if (!contentType && typeof payload === 'string') {
     reply.res.setHeader('Content-Type', 'text/plain')
+  } else if (reply._serializer) {
+    payload = reply._serializer(payload)
   } else if (!contentType || contentType === 'application/json') {
     reply.res.setHeader('Content-Type', 'application/json')
-    // Here we are assuming that the custom serializer is a json serializer
-    payload = reply._serializer ? reply._serializer(payload) : serialize(reply.context, payload, reply.res.statusCode)
-    flatstr(payload)
-  } else if (reply._serializer) { // All the code below must have a 'content-type' setted
-    payload = reply._serializer(payload)
+    payload = serialize(reply.context, payload, reply.res.statusCode)
   }
 
+  flatstr(payload)
```

Comment:
This is not correct. `payload` could be a buffer be a buffer here. In the previous code this check was with a `application/json` content type, for which we have a fast case. I do not know what happens passing a `Buffer`  through `Number()`, but I would avoid it anyway.

Can you move `flatstr` within the previous else-if as it was before?